File: eval-reviews/2026-09-12-chained/query-null/typescript/fable/r1/starter/engine.ts
/** Stable relational operators and real constant-fold / scan-filter rewrites. */
import { aggregates } from "./model.ts";
import type { Expr, Plan, ScalarType, Value } from "./model.ts";
import { walk } from "./binder.ts";
type Known = Exclude<Value, null>;
/**
 * Three-valued evaluation. Booleans are represented directly: TRUE is `true`,
 * FALSE is `false`, and UNKNOWN is `null`, the same value as SQL NULL.
 */
export function evaluate(e: Expr, row: Value[], group: Value[][] = []): Value {
  switch (e.op) {
    case "lit":
      return e.value as Value;
    case "col":
      return row[e.index!];
    case "COUNT":
      return e.args.length
        ? group.filter((r) => evaluate(e.args[0], r) !== null).length
        : group.length;
    case "SUM":
    case "MIN":
    case "MAX": {
      const vs = group
        .map((r) => evaluate(e.args[0], r))
        .filter((v): v is Known => v !== null);
      if (!vs.length) return null;
      if (e.op === "SUM") return vs.reduce<number>((n, v) => n + (v as number), 0);
      return vs.reduce((a, b) => (e.op === "MIN" ? (b < a ? b : a) : b > a ? b : a));
    }
    case "COALESCE":
      for (const a of e.args) {
        const v = evaluate(a, row, group);
        if (v !== null) return v;
      }
      return null;
    case "IS NULL":
      return evaluate(e.args[0], row, group) === null;
    case "IS NOT NULL":
      return evaluate(e.args[0], row, group) !== null;
    case "NOT": {
      const a = evaluate(e.args[0], row, group);
      return a === null ? null : !a;
    }
    case "AND": {
      const a = evaluate(e.args[0], row, group);
      const b = evaluate(e.args[1], row, group);
      if (a === false || b === false) return false;
      return a === null || b === null ? null : true;
    }
    case "OR": {
      const a = evaluate(e.args[0], row, group);
      const b = evaluate(e.args[1], row, group);
      if (a === true || b === true) return true;
      return a === null || b === null ? null : false;
    }
  }
  const a = evaluate(e.args[0], row, group);
  const b = evaluate(e.args[1], row, group);
  if (a === null || b === null) return null;
  switch (e.op) {
    case "+":
      return (a as number) + (b as number);
    case "-":
      return (a as number) - (b as number);
    case "*":
      return (a as number) * (b as number);
    case "=":
      return a === b;
    case "<>":
      return a !== b;
    case "<":
      return a < b;
    case "<=":
      return a <= b;
    case ">":
      return a > b;
    case ">=":
      return a >= b;
    default:
      throw new Error(`Unknown bound operator ${e.op}`);
  }
}
function literal(value: Value, type?: ScalarType): Expr {
  return { op: "lit", value, args: [], type, nullable: value === null };
}
function isLiteral(e: Expr, value: Value): boolean {
  return e.op === "lit" && e.value === value;
}
/** Structural equality of bound, aggregate-free deterministic expressions. */
function same(a: Expr, b: Expr): boolean {
  return (
    a.op === b.op &&
    a.index === b.index &&
    JSON.stringify(a.value ?? null) === JSON.stringify(b.value ?? null) &&
    a.args.length === b.args.length &&
    a.args.every((x, i) => same(x, b.args[i]))
  );
}
/** Operators that yield NULL whenever any operand is NULL. */
const strict = new Set(["+", "-", "*", "=", "<>", "<", "<=", ">", ">=", "NOT"]);
/**
 * Bottom-up simplification that is sound under three-valued logic. Rewrites
 * that would be tautologies in two-valued logic (`x = x`, `p OR NOT p`,
 * `x IS NOT NULL`) fire only when the binder proved the operand non-nullable.
 */
function fold(e: Expr): Expr {
  e.args = e.args.map(fold);
  if (aggregates.has(e.op) || e.op === "lit" || e.op === "col") return e;
  if (e.op === "COALESCE") {
    // Drop NULL literals; nothing after the first non-nullable argument matters.
    const kept: Expr[] = [];
    for (const a of e.args) {
      if (isLiteral(a, null)) continue;
      kept.push(a);
      if (!a.nullable) break;
    }
    if (kept.length === 0) return literal(null, e.type);
    if (kept.length === 1 && kept[0].type === e.type) return kept[0];
    e.args = kept;
  }
  if (e.args.every((a) => a.op === "lit")) return literal(evaluate(e, []), e.type);
  const [a, b] = e.args;
  if (strict.has(e.op) && e.args.some((x) => isLiteral(x, null)))
    return literal(null, e.type);
  if (e.op === "NOT" && a.op === "NOT") return a.args[0];
  if (e.op === "AND" || e.op === "OR") {
    // FALSE absorbs AND and TRUE absorbs OR even against UNKNOWN; the other
    // constant is an identity. NULL literals cannot be removed.
    const absorbing = e.op === "OR";
    if (e.args.some((x) => isLiteral(x, absorbing))) return literal(absorbing, "bool");
    if (isLiteral(a, !absorbing)) return b;
    if (isLiteral(b, !absorbing)) return a;
    const complementary =
      (b.op === "NOT" && same(a, b.args[0])) || (a.op === "NOT" && same(a.args[0], b));
    if (complementary && !a.nullable) return literal(absorbing, "bool");
  }
  if (e.op === "=" && !a.nullable && same(a, b)) return literal(true, "bool");
  if (e.op === "IS NULL" && !a.nullable) return literal(false, "bool");
  if (e.op === "IS NOT NULL" && !a.nullable) return literal(true, "bool");
  return e;
}
function conjuncts(e: Expr): Expr[] {
  return e.op === "AND" ? e.args.flatMap(conjuncts) : [e];
}
export function optimize(plan: Plan): Plan {
  const q = plan.query;
  q.select = q.select.map(([e, a]) => [fold(e), a]);
  q.joins = q.joins.map(fold);
  if (q.having) q.having = fold(q.having);
  if (q.where) {
    const width = plan.tables.reduce((n, t) => n + t.columns.length, 0);
    const remaining: Expr[] = [];
    for (const e of conjuncts(fold(q.where))) {
      if (isLiteral(e, true)) continue;
      const sources = new Set(
        [...walk(e)].filter((n) => n.op === "col").map((n) => n.source!),
      );
      if (plan.tables.length > 1 && sources.size === 1) {
        const s = [...sources][0];
        if (s > 0 && q.outer[s - 1]) {
          // The source may be NULL-padded. A conjunct over only its columns can
          // move below the join solely when it rejects the all-NULL padding, in
          // which case the LEFT JOIN is equivalent to an INNER JOIN.
          if (evaluate(e, Array<Value>(width).fill(null)) === true) {
            remaining.push(e);
            continue;
          }
          q.outer[s - 1] = false;
        }
        plan.filters[s].push(e);
      } else remaining.push(e);
    }
    q.where = remaining.reduce<Expr | undefined>(
      (a, e) => (a ? { op: "AND", args: [a, e], type: "bool" } : e),
      undefined,
    );
  }
  return plan;
}
export function execute(plan: Plan): { columns: string[]; rows: Value[][] } {
  const q = plan.query;
  let offset = 0;
  const scans = plan.tables.map((t, i) => {
    const rows = t.rows.filter((r) =>
      plan.filters[i].every(
        (p) => evaluate(p, [...Array<Value>(offset), ...r]) === true,
      ),
    );
    offset += t.columns.length;
    return rows;
  });
  let rows = scans[0];
  q.joins.forEach((on, i) => {
    const padding = Array<Value>(plan.tables[i + 1].columns.length).fill(null);
    rows = rows.flatMap((l) => {
      const matched = scans[i + 1]
        .map((r) => [...l, ...r])
        .filter((r) => evaluate(on, r) === true);
      return matched.length || !q.outer[i] ? matched : [[...l, ...padding]];
    });
  });
  if (q.where) rows = rows.filter((r) => evaluate(q.where!, r) === true);
  let units: [Value[], Value[][]][];
  if (plan.aggregate) {
    const groups = new Map<string, Value[][]>();
    for (const row of rows) {
      const key = JSON.stringify(q.groups.map((e) => evaluate(e, row)));
      const g = groups.get(key);
      if (g) g.push(row);
      else groups.set(key, [row]);
    }
    if (!q.groups.length && !rows.length) groups.set("[]", []);
    units = [...groups.values()].map((g) => [g[0] ?? [], g]);
  } else units = rows.map((r) => [r, []]);
  let projected = units
    .filter(([r, g]) => !q.having || evaluate(q.having, r, g) === true)
    .map(([r, g]) => q.select.map(([e]) => evaluate(e, r, g)));
  if (q.distinct) {
    const seen = new Set<string>();
    projected = projected.filter((r) => {
      const key = JSON.stringify(r);
      if (seen.has(key)) return false;
      seen.add(key);
      return true;
    });
  }
  projected.sort((a, b) => {
    for (const [index, desc, nullsFirst] of q.order) {
      const x = a[index as number],
        y = b[index as number];
      if (x === null || y === null) {
        if (x === y) continue;
        return (x === null) === nullsFirst ? -1 : 1;
      }
      const c = x < y ? -1 : x > y ? 1 : 0;
      if (c) return desc ? -c : c;
    }
    return 0;
  });
  return {
    columns: q.select.map(([, a]) => a),
    rows: projected.slice(
      q.offset,
      q.limit === undefined ? undefined : q.offset + q.limit,
    ),
  };
}

File: eval-reviews/2026-09-12-chained/query-null/python/fable/r1/starter/engine.py
"""Stable relational execution and semantics-preserving optimization.

Values are Python int/str/bool, with SQL NULL represented as None. Predicates
follow three-valued logic: True, False, or None (UNKNOWN); only True passes a
WHERE, ON, or HAVING filter."""
from model import Expr
from parser import AGGREGATES
from binder import walk


def is_true(v):
    return v is True


def evaluate(e, row, group=None):
    """Evaluate `e` on a flattened row. For aggregate queries `group` is either
    the list of member rows or, for incrementally maintained views, a mapping
    from the aggregate node's id() to its already-maintained value."""
    if e.op == 'lit': return e.value
    if e.op == 'col': return row[e.index]
    if e.op in AGGREGATES:
        if isinstance(group, dict): return group[id(e)]
        if e.op == 'COUNT' and not e.args: return len(group)
        vs = [v for v in (evaluate(e.args[0], r) for r in group) if v is not None]
        if e.op == 'COUNT': return len(vs)
        if not vs: return None
        return {'SUM': sum, 'MIN': min, 'MAX': max}[e.op](vs)
    if e.op == 'COALESCE':
        for a in e.args:
            v = evaluate(a, row, group)
            if v is not None: return v
        return None
    a = evaluate(e.args[0], row, group)
    if e.op == 'ISNULL': return a is None
    if e.op == 'ISNOTNULL': return a is not None
    if e.op == 'NOT': return None if a is None else not a
    b = evaluate(e.args[1], row, group)
    if e.op == 'AND':
        if a is False or b is False: return False
        return None if a is None or b is None else True
    if e.op == 'OR':
        if a is True or b is True: return True
        return None if a is None or b is None else False
    if a is None or b is None: return None
    return SCALAR[e.op](a, b)


SCALAR = {
    '+': lambda a, b: a + b,
    '-': lambda a, b: a - b,
    '*': lambda a, b: a * b,
    '=': lambda a, b: a == b,
    '<>': lambda a, b: a != b,
    '<': lambda a, b: a < b,
    '<=': lambda a, b: a <= b,
    '>': lambda a, b: a > b,
    '>=': lambda a, b: a >= b,
}


def literal(value, type):
    return Expr('lit', value, type=type, nullable=value is None)


def same_column(a, b):
    return a.op == 'col' and b.op == 'col' and a.index == b.index


def fold(e):
    """Bottom-up constant folding that respects three-valued logic.

    Rewrites that are only valid for non-null operands (x = x, x IS NULL) use
    the binder's nullable flag, which already accounts for outer-join padding."""
    e.args = [fold(a) for a in e.args]
    if e.op in AGGREGATES | {'lit', 'col'}: return e
    if all(a.op == 'lit' for a in e.args):
        return literal(evaluate(e, []), e.type)
    if e.op in ('AND', 'OR'):
        absorbing, identity = (False, True) if e.op == 'AND' else (True, False)
        for a in e.args:
            if a.op == 'lit' and a.value is absorbing: return literal(absorbing, 'bool')
        for a, other in ((e.args[0], e.args[1]), (e.args[1], e.args[0])):
            if a.op == 'lit' and a.value is identity: return other
    elif e.op == 'ISNULL' and not e.args[0].nullable: return literal(False, 'bool')
    elif e.op == 'ISNOTNULL' and not e.args[0].nullable: return literal(True, 'bool')
    elif e.op in ('=', '<=', '>=') and same_column(*e.args) and not e.args[0].nullable: return literal(True, 'bool')
    elif e.op in ('<>', '<', '>') and same_column(*e.args) and not e.args[0].nullable: return literal(False, 'bool')
    elif e.op == 'COALESCE':
        args = []
        for a in e.args:
            if a.op == 'lit' and a.value is None: continue
            args.append(a)
            if not a.nullable: break
        if not args: return literal(None, e.type)
        if len(args) == 1: return args[0]
        e.args = args
    return e


def conjuncts(e):
    if e.op == 'AND': return conjuncts(e.args[0]) + conjuncts(e.args[1])
    return [e]


def null_rejecting(e, width):
    """True when the conjunct (referencing only `source`) cannot be TRUE on a
    row padded with NULL for that source, i.e. it discards outer-join padding."""
    return not is_true(evaluate(e, [None] * width))


def optimize(plan):
    q = plan.query
    q.select = [(fold(e), a) for e, a in q.select]
    q.joins = [fold(e) for e in q.joins]
    if q.having: q.having = fold(q.having)
    if q.where:
        width = sum(len(t['columns']) for t in plan.tables)
        remaining = []
        for e in conjuncts(fold(q.where)):
            if e.op == 'lit' and e.value is True: continue
            remaining.append(e)
        # A filter on a single source can run at that source's scan when every
        # output row carries that source's real values (it is never padded), or
        # when the filter rejects padded rows, in which case the LEFT JOIN that
        # pads it is equivalent to an INNER JOIN and can be converted first.
        changed = True
        while changed:
            changed, kept = False, []
            for e in remaining:
                sources = {n.source for n in walk(e) if n.op == 'col'}
                if len(plan.tables) > 1 and len(sources) == 1:
                    s = next(iter(sources))
                    padded = s > 0 and q.kinds[s - 1] == 'LEFT'
                    if padded and null_rejecting(e, width):
                        q.kinds[s - 1] = 'INNER'
                        padded = False
                    if not padded:
                        plan.filters[s].append(e)
                        changed = True
                        continue
                kept.append(e)
            remaining = kept
        q.where = None
        for e in remaining:
            q.where = e if q.where is None else Expr('AND', args=[q.where, e], type='bool')
    return plan


def sort_key(index, descending, nulls_first):
    # Python's sort is stable for reverse=True as well; the flag orders NULLs
    # relative to values and the tuple length difference avoids comparing None.
    null_rank = 0 if nulls_first != descending else 1
    return lambda r: (null_rank,) if r[index] is None else (1 - null_rank, r[index])


def execute(plan):
    q = plan.query
    scans, offset = [], 0
    for t, predicates in zip(plan.tables, plan.filters):
        scans.append([r for r in t['rows'] if all(is_true(evaluate(p, [None] * offset + r)) for p in predicates)])
        offset += len(t['columns'])
    rows = scans[0]
    for right, on, kind, t in zip(scans[1:], q.joins, q.kinds, plan.tables[1:]):
        padding = [None] * len(t['columns'])
        joined = []
        for l in rows:
            matches = [l + r for r in right if is_true(evaluate(on, l + r))]
            if not matches and kind == 'LEFT': matches = [l + padding]
            joined.extend(matches)
        rows = joined
    if q.where: rows = [r for r in rows if is_true(evaluate(q.where, r))]
    if plan.aggregate:
        groups = {}
        for row in rows:
            key = tuple(evaluate(e, row) for e in q.groups)
            groups.setdefault(key, []).append(row)
        if not q.groups and not rows: groups[()] = []
        units = [(g[0] if g else [], g) for g in groups.values()]
    else: units = [(r, None) for r in rows]
    projected = [[evaluate(e, r, g) for e, _ in q.select] for r, g in units if q.having is None or is_true(evaluate(q.having, r, g))]
    if q.distinct: projected = [list(r) for r in dict.fromkeys(tuple(r) for r in projected)]
    for i, descending, nulls_first in reversed(q.order): projected.sort(key=sort_key(i, descending, nulls_first), reverse=descending)
    projected = projected[q.offset: None if q.limit is None else q.offset + q.limit]
    return {'columns': [a for _, a in q.select], 'rows': projected}
